#!/bin/sh

#VCF files contain training sets generated according to criteria described in the Methods section. This command produces the BAM annotations necessary to apply VQSR
java -Xmx10g -jar /mnt/PDanalysis/scripts/GenomeAnalysisTK-2.3-9-ge5ebf34/GenomeAnalysisTK.jar -T VariantRecalibrator -R /mnt/PDanalysis/results/trinity_assemblies/Trinity_kmer31.fasta -input /mnt/PDanalysis/process/GATK_kmer31-SE/raw.snps.annotated.complete.vcf -resource:good,known=false,training=true,truth=true,prior=15.0 /mnt/PDanalysis/process/GATK_kmer31-SE/good.snps.recode.vcf -resource:bad,known=false,training=true,truth=false,bad=true,prior=0.0 /mnt/PDanalysis/process/GATK_kmer31-SE/bad.snps.recode.vcf --percentBadVariants 0.05 -mode SNP -recalFile /mnt/PDanalysis/process/GATK_kmer31-SE/output.complete.recal -tranchesFile /mnt/PDanalysis/process/GATK_kmer31-SE/output.complete.tranches -rscriptFile /mnt/PDanalysis/process/GATK_kmer31-SE/output.complete.plots.R --target_titv 1.34 --ts_filter_level 90.0 --maxGaussians 4 -an ABHom -an ABHet -an BaseQRankSum -an DP -an HaplotypeScore -an InbreedingCoeff -an MQ -an MQRankSum -an QD -an ReadPosRankSum > /mnt/PDanalysis/process/GATK_kmer31-SE/VariantRecalibrator.complete.log

#apply VQSR, filtering at 90% truth sensitivity threshold
java -jar /mnt/PDanalysis/scripts/GenomeAnalysisTK-2.3-9-ge5ebf34/GenomeAnalysisTK.jar -R /mnt/PDanalysis/results/trinity_assemblies/Trinity_kmer31.fasta -T ApplyRecalibration -input raw.snps.vcf --ts_filter_level 90.0 -tranchesFile VQSR-90/output.complete.tranches -recalFile VQSR-90/output.complete.recal -mode SNP -o recalibrated.snps.complete.filtered.vcf > ApplyRecalibration.log
